// allure-deployer-shared/dist/features/messaging/slack-notifier.js
import { appLog } from "../../utilities/util.js";
export class SlackNotifier {
    constructor(client, args) {
        this.slackClient = client;
        this.args = args;
    }
    buildStatusBlock(label, emoji, count) {
        return {
            type: "context",
            elements: [
                {
                    type: "mrkdwn",
                    text: `${emoji}  *${label}:* ${count}`,
                },
            ],
        };
    }
    buildButtonBlock(text, url) {
        return {
            type: "actions",
            elements: [
                {
                    type: "button",
                    text: {
                        type: "plain_text",
                        text,
                        emoji: true,
                    },
                    url,
                },
            ],
        };
    }
    async notify(data) {
        const blocks = [
            {
                type: "section",
                text: {
                    type: "mrkdwn",
                    text: "*Your Allure report is ready* 📊",
                },
            },
        ];
        // Add status blocks
        const { resultStatus } = data;
        if (resultStatus.passed)
            blocks.push(this.buildStatusBlock("Passed", ":white_check_mark:", resultStatus.passed));
        if (resultStatus.broken)
            blocks.push(this.buildStatusBlock("Broken", ":warning:", resultStatus.broken));
        if (resultStatus.skipped)
            blocks.push(this.buildStatusBlock("Skipped", ":next_track_button:", resultStatus.skipped));
        if (resultStatus.failed)
            blocks.push(this.buildStatusBlock("Failed", ":x:", resultStatus.failed));
        if (resultStatus.unknown)
            blocks.push(this.buildStatusBlock("Unknown", ":question:", resultStatus.unknown));
        // Add report and storage buttons
        if (data.reportUrl)
            blocks.push(this.buildButtonBlock("View report", data.reportUrl));
        if (data.storageUrl)
            blocks.push(this.buildButtonBlock("View files in storage", data.storageUrl));
        // Add GitHub promotion button
        blocks.push(this.buildButtonBlock("Star us on GitHub :smile:", "https://github.com/cybersokari/allure-report-deployer"));
        try {
            await this.slackClient.postMessage(blocks, 'Your Allure report is ready.');
            appLog('Slack message sent');
        }
        catch (error) {
            console.error('Error sending Slack message:', error);
        }
    }
}
